chia/full_node/subscriptions.py

```diff
@@ -1,144 +1,179 @@
-            self._peer_sub_counter[peer_id] += 1
-            subscriptions_left -= 1
-
-            if subscriptions_left == 0:
-                log.info(
-                    "peer_id: %s reached max number of puzzle-hash subscriptions. "
-                    "Not all its coin states will be reported",
-                    peer_id,
-                )
-                break
-        return ret
+        assert row is not None
+
+        return int(row[0]) > 0
+
+    def is_coin_subscribed(self, coin_id: bytes32) -> bool:
+        with self._db_conn as conn:
+            row = conn.execute("SELECT COUNT(*) FROM coin_subscriptions WHERE coin_id=?", (coin_id,)).fetchone()
+
+        assert row is not None
+
+        return int(row[0]) > 0
 
-    def add_coin_subscriptions(self, peer_id: bytes32, coin_ids: List[bytes32], max_items: int) -> None:
-        coin_id_peers = self._peer_coin_ids.setdefault(peer_id, set())
-        existing_sub_count = self._peer_sub_counter.setdefault(peer_id, 0)
+    def _add_subscriptions(
+        self, peer_id: bytes32, table_name: str, item_name: str, items: List[bytes32], max_items: int
+    ) -> Set[bytes32]:
+        existing_sub_count = self.peer_subscription_count(peer_id)
+        inserted: Set[bytes32] = set()
 
-        # if we've reached the limit on number of subscriptions, just bail
+        # If we've reached the subscription limit, just bail.
         if existing_sub_count >= max_items:
             log.info(
-                "peer_id: %s reached max number of coin subscriptions. Not all its coin states will be reported",
+                "Peer %s reached the subscription limit. Not all coin states will be reported.",
                 peer_id,
             )
-            return
-
-        # decrement this counter as we go, to know if we've hit the limit of
-        # number of subscriptions
-        subscriptions_left = max_items - existing_sub_count
-
-        for coin_id in coin_ids:
-            coin_sub = self._coin_subscriptions.setdefault(coin_id, set())
-            if peer_id in coin_sub:
-                continue
-
-            coin_sub.add(peer_id)
-            coin_id_peers.add(coin_id)
-            self._peer_sub_counter[peer_id] += 1
-            subscriptions_left -= 1
-
-            if subscriptions_left == 0:
-                log.info(
-                    "peer_id: %s reached max number of coin subscriptions. Not all its coin states will be reported",
-                    peer_id,
-                )
-                break
+            return inserted
+
+        with self._db_conn as conn:
+            # Decrement this counter as we go, to know if we've hit the subscription limit.
+            subscriptions_left = max_items - existing_sub_count
+
+            for item in items:
+                row = conn.execute(
+                    f"SELECT COUNT(*) FROM {table_name} WHERE peer_id=? AND {item_name}=?",
+                    (peer_id, item),
+                ).fetchone()
```

Comment:
it looks really odd to count the number of rows here just to decide whether we try to insert it or not.
I would expect it to be a lot faster to just insert it and handle the error if it's already added.